File: 2025_12/leetcode2025-12-13.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        string chars = "_0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        unordered_map<string, int> mp;
        mp["electronics"] = 0;
        mp["grocery"] = 1;
        mp["pharmacy"] = 2;
        mp["restaurant"] = 3;
        vector<vector<string>> tot(4);
        for (int i=0; i<code.size(); i++){
            string curCode = code[i], b = businessLine[i];
            bool flag = true;
            for (auto& c : curCode){
                if (count(chars.begin(), chars.end(), c) == 0){
                    flag = false;
                    break;
                }
            }
            if (curCode.size()>0 && flag && mp.count(b) && isActive[i]){
                tot[mp[b]].push_back(curCode);
            }
        }
        for (int i=0; i<4; i++){
            sort(tot[i].begin(), tot[i].end());
        }
        vector<string> res;
        for (int i=0; i<4; i++){
            for (auto& s : tot[i]){
                res.push_back(s);
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<string> code = {"SAVE20","","PHARMA5","SAVE@20"}, businessLine = {"restaurant","grocery","pharmacy","restaurant"};
    vector<bool> isActive = {true,true,true,true};
    vector<string> res = s.validateCoupons(code, businessLine, isActive);
    for (auto& i : res){
        cout << i << " ";
    }
    cout << endl;
}

File: leetcode2025/2025_12/leetcode2025-12-13.py
from typing import List
from string import *

class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        n = len(code)
        dic = {"electronics":0, "grocery":1, "pharmacy":2, "restaurant":3}
        chars = '0123456789'+ascii_lowercase+ascii_uppercase+'_'
        res = []
        for i in range(n):
            cod, b, ia = code[i], businessLine[i], isActive[i]
            if len(cod)>0 and all(c in chars for c in cod) and b in dic.keys() and ia:
                res.append((dic[b], cod))
        res.sort()
        return [i[1] for i in res]

if __name__ == "__main__":
    s = Solution()
    code = ["SAVE20","","PHARMA5","SAVE@20"]
    businessLine = ["restaurant","grocery","pharmacy","restaurant"]
    isActive = [True,True,True,True]
    print(s.validateCoupons(code, businessLine, isActive))
